#166272334 Verify User Account

emmygozi opened the pull request:
#### What does this PR do?

This pull request verifies a new user on signup or on login.
It caters for scenarios when verification link expires by creating two new APIs to generate verification link and verify a user respectively.
#### Description of Task to be completed?

- create API on sendgrid
- send verification mail on signup
- API to generate verification mail in case of link expiration
- create verify mail model and migration
- verifies a new user
- Sync with Macco to use mail generating template
- Write unit tests to improve test coverage and test uncovered files
#### How should this be manually tested?

#### Any background context you want to provide?

Had to work in tandem with Macco to agree on templates to use
#### What are the relevant pivotal tracker stories?

https://www.pivotaltracker.com/story/show/166272334
#### Screenshots (if appropriate)

![image](https://user-images.githubusercontent.com/32358797/59857824-a4623d80-9371-11e9-84d8-341787684390.png)

#### Questions:

N/A

emmygozi committed:
feat-[166272340]: user create article

-user should be able to create an article
-article should have a unique slug
-article is time stamped
-article can have images

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- fix bug in user model
- add new column image to article
- fix hound exception

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- write test for articles controller
- crud poerations for articles

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- write test for articles controller
- crud operations for articles

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- write test for articles controller
- crud operations for articles

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- write test for articles controller
- crud operations for articles

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Article should also return ratings and authors detail
- write test for models

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Refactor code
- article should also return user rating and author details

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Refactor code
- article should also return user rating and author details

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Refactor code
- article should also return user rating and author details

[Delivers #166272340]

emmygozi committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272340-ft-user-create-article

emmygozi committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272340-ft-user-create-article

emmygozi committed:
feat-[166272340]: user create article

- Refactor code
- article should also return user rating and author details

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement LF feedback
- rename instance of decoded to user in articles controller

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement LF feedback
- rename instance of decoded to user in articles controller

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement LF feedback
- rename instance of decoded to user in articles controller

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Trim title
- Add image to article payload

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement feedback to use user instance from database in articles update

[Delivers #166272340]

emmygozi committed:
feat-[166272340]: user create article

- Implement feedback to use user instance from database in articles update

[Delivers #166272340]

emmygozi committed:
chore-[166751405]: write unit tests

- Create tests to cover all lines in articles controller
- Create tests to cover all lines in profiles controller
- Creates unit tests for users controller
- Tests helper fuctions Token and articles payload

emmygozi committed:
chore-[166751405]: write unit tests

- Create tests to cover all lines in articles controller
- Create tests to cover all lines in profiles controller
- Creates unit tests for users controller
- Tests helper fuctions Token and articles payload

emmygozi committed:
chore-[166751405]: write unit tests

- Create tests to cover all lines in articles controller
- Create tests to cover all lines in profiles controller
- Creates unit tests for users controller
- Tests helper fuctions Token and articles payload

emmygozi committed:
chore-[166751405]: write unit tests

- Create tests to cover all lines in articles controller
- Create tests to cover all lines in profiles controller
- Creates unit tests for users controller
- Tests helper fuctions Token and articles payload

emmygozi committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166751405-ch-write-unit-tests

emmygozi committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272334-ft-verification-mail-signup

emmygozi committed:
feat-[166272334]: verify email on signup

- create api on sendgrid
- send verification mail on signup
- api to generate verification mail in case of link expiration
- create verify mail model and migration
- verifies a new user

[Finishes #166272334]

emmygozi committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272334-ft-verification-mail-signup

emmygozi committed:
feat-[166272334]: email verification

- create api on sendgrid
- send verification mail on signup
- generation of verification mail
- create verify mail model and migration
- verifies a new user

[Finishes #166272334]

emmygozi committed:
Merge branch '166272334-ft-verification-mail-signup' of https://github.com/andela/kingsmen-ah-backend into 166272334-ft-verification-mail-signup

emmygozi committed:
feat-[166272334]: email verification

- create api on sendgrid
- send verification mail on signup
- generation of verification mail
- create verify mail model and migration
- verifies a new user

[Finishes #166272334]

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -37,7 +43,19 @@ class UserController {
+      await VerifyUser.create({ ...verifyDetails });
+      sendVerifyMailToken(verifyToken, user.email, user.username);
+      return res.status(201).json({
+        status: 'success', message: 'User created successfully', user: userExtractor(user, token), verifyToken
```
why exactly are you sending the `verifyToken` ?

emmygozi replied:
For my test to have access to the verification token.

fidelisojeah replied:
Remove it.

Find another way to write your tests

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -152,6 +169,91 @@ class UserController {
+    try {
+      const { token, email } = req.query;
+
+      if (!token || !email) return Response.error(res, 400, 'Please use a valid reset link');
```
reset?

emmygozi replied:
Would fix this. Typo.

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -152,6 +169,91 @@ class UserController {
+
+      if (!token || !email) return Response.error(res, 400, 'Please use a valid reset link');
+      const user = await User.findOne({ where: { email } });
+      if (!user) return Response.error(res, 404, 'User does not exist');
```
lol what?

firstly, this wouldn't exactly be a 404 (considering the username is not in the endpoint)
Also, you are releasing too much information by doing this.
`Invalid Token` would work here

emmygozi replied:
Alright on it.

fidelisojeah commented on server/helpers/mailer.js:
```diff
@@ -0,0 +1,57 @@
+
+config();
+
+const url = process.env.UI_HOST_NAME || 'https://kingsmen-ah-frontend-staging.herokuapp.com';
```
nah

you would be better off defaulting this to localhost.

fidelisojeah reviewed (changes requested):
just minor changes

emmygozi committed:
feat-[166272334]: email verification

- create api on sendgrid
- send verification mail on signup
- generation of verification mail
- create verify mail model and migration
- verifies a new user

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: email verification

- create api on sendgrid
- send verification mail on signup
- generation of verification mail
- create verify mail model and migration
- verifies a new user

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: email verification

- create api on sendgrid
- send verification mail on signup
- generation of verification mail
- create verify mail model and migration
- verifies a new user

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: email verification

- update test

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: email verification

- update test

[Finishes #166272334]

emmygozi committed:
ft-[166272334]: user verification

- Implement feedback

[Finishes #166272334]

emmygozi committed:
ft-[166272334]: user verification

- Implement feedback

[Finishes #166272334]

emmygozi committed:
ft-[166272334]: user verification

- Implement feedback

[Finishes #166272334]

emmygozi committed:
ft-[166272334]: user verification

- Implement feedback

[Finishes #166272334]

emmygozi committed:
ft-[166272334]: verify user account

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: verify user

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: verify user

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: verify user

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: verify user

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: verify user

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: verify user

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
Merge branch '166272334-ft-verification-mail-signup' of https://github.com/andela/kingsmen-ah-backend into 166272334-ft-verification-mail-signup

emmygozi committed:
feat-[166272334]: Verify user

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: Verify user

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: Verify user

- implement feedback
- user account verification

[Finishes #166272334]

emmygozi committed:
feat-[166272334]: Verify user

- Add swagger documentation
- user account verification

[Finishes #166272334]

GerrardE reviewed (approved):
Good One
